train.py

```diff
@@ -90,29 +90,49 @@ def load_state_dict(self, state_dict) -> None:
         self.log_steps = torch.load(state_dict["log_steps"], weights_only=False)
 
 
-def build_optimizer(model, job_config: JobConfig):
-    # build optimizer
-    name = job_config.optimizer.name
-    lr = job_config.optimizer.lr
-    fused = job_config.optimizer.fused
-
-    # Common parameters for both optimizers
-    optimizer_kwargs = {
-        "lr": lr,
-        "betas": (0.9, 0.95),
-        "weight_decay": 0.1,
-        "fused": fused,
-        "foreach": not fused,
-    }
-    if name == "Adam":
-        # TODO: make the optimizer options configurable by toml/cmd args
-        optimizer = torch.optim.Adam(model.parameters(), **optimizer_kwargs)
-    elif name == "AdamW":
-        optimizer = torch.optim.AdamW(model.parameters(), **optimizer_kwargs)
-    else:
-        raise NotImplementedError(f"Optimizer {name} not added.")
+def build_optimizers(model_parts, job_config: JobConfig):
+    """Wrap one optimizer per model part in an OptimizersContainer which provides a single
+    step() and zero_grad() method for all the child optimizers.
+    """
+
+    def _build_optimizer(model):
+        name = job_config.optimizer.name
+        lr = job_config.optimizer.lr
+        fused = job_config.optimizer.fused
+
+        # Common parameters for both optimizers
+        optimizer_kwargs = {
+            "lr": lr,
+            "betas": (0.9, 0.95),
+            "weight_decay": 0.1,
+            "fused": fused,
+            "foreach": not fused,
+        }
+        if name == "Adam":
+            # TODO: make the optimizer options configurable by toml/cmd args
+            optimizer = torch.optim.Adam(model.parameters(), **optimizer_kwargs)
+        elif name == "AdamW":
+            optimizer = torch.optim.AdamW(model.parameters(), **optimizer_kwargs)
+        else:
+            raise NotImplementedError(f"Optimizer {name} not added.")
+
+        return optimizer
+
+    class OptimizersContainer:
```

Comment:
Since we are going to have a class anyway, will it be better to have a `Optimizers` class to do everything, including the state_dict part?

```
class Optimizers:
    def __init__(self, model_parts, job_config: JobConfig) -> None:
            original build optimizer code
            state_dict init code
            
    def step(self):
             ...
             
     def  state_dict(self) -> None:
             OptimizerWrapper state_dict code
             
      def load_state_dict(self) -> None:
             OptimizerWrapper load_state_dict code
``` 

And can we also do this for model? It seems that the main training loop will need to loop all the model parts as well. Having a Models and a Optimizer classes can encapsulate the logic and remove extra classes (ModelWrapper and OptimizerWrapper).

Reply:
I am going to land this PR first.  Then make a new PR to try some refactoring including this suggestion, see if we can iterate to something we like.